File: app/risk/market_clock.py
"""IST market-session clock (RISK.md §1, §5).

Pure helpers over `datetime` that answer three questions the risk
layer needs:

  - `is_market_open(now)`    — are we inside the trading session at all?
  - `is_entry_window(now)`   — may we OPEN a new position now? (excludes
                               the first 15 min after open and the last
                               30 min before close — the noisy edges)
  - `square_off_due(now)`    — is it time to force-flatten intraday
                               positions? (≥ SQUARE_OFF_TIME_IST)

All times come from `Settings` ("HH:MM" IST strings) and are read live,
so an operator can shift the windows without a restart. `now` is an
optional timezone-aware (or naive-UTC) datetime, injectable for tests;
it defaults to the real wall clock.

NOTE (seam): there is no exchange-holiday calendar wired in — these
functions only exclude weekends. A holiday feed can be added later by
extending `_is_trading_day`. Until then, treat a holiday like any open
weekday; the EOD square-off and the broker's own rejection of orders on
a closed exchange are the backstops.
"""
from __future__ import annotations

from datetime import datetime, time, timedelta, timezone
from typing import Optional

from app.config import get_settings

# India Standard Time is a fixed UTC+5:30 offset (no DST).
IST = timezone(timedelta(hours=5, minutes=30))


def _parse_hhmm(value: str) -> time:
    """Parse a validated "HH:MM" string into a `time`. Settings already
    validates the format, so this is a thin, defensive split."""
    hh, mm = value.split(":")
    return time(hour=int(hh), minute=int(mm))


def to_ist(now: Optional[datetime] = None) -> datetime:
    """Return `now` as an IST-aware datetime.

    A naive datetime is assumed to be UTC (the project-wide convention),
    matching how `filed_at` and friends are stored.
    """
    if now is None:
        now = datetime.now(timezone.utc)
    if now.tzinfo is None:
        now = now.replace(tzinfo=timezone.utc)
    return now.astimezone(IST)


def _is_trading_day(ist_dt: datetime) -> bool:
    # Monday=0 .. Friday=4 are trading days. Holidays are not modelled
    # (see module docstring).
    return ist_dt.weekday() < 5


def is_market_open(now: Optional[datetime] = None) -> bool:
    """True iff `now` (IST) is a weekday within [MARKET_OPEN, MARKET_CLOSE]."""
    s = get_settings()
    ist = to_ist(now)
    if not _is_trading_day(ist):
        return False
    t = ist.time()
    return _parse_hhmm(s.MARKET_OPEN_IST) <= t <= _parse_hhmm(s.MARKET_CLOSE_IST)


def is_entry_window(now: Optional[datetime] = None) -> bool:
    """True iff a NEW position may be opened now.

    The entry window is a weekday inside [ENTRY_WINDOW_START,
    ENTRY_WINDOW_END] — by default 09:30–15:00 IST, i.e. the session
    minus the first 15 min (opening auction noise) and the last 30 min
    (pre-close mean reversion + square-off run-up).

    When `Settings.ENFORCE_MARKET_HOURS` is False the gate is disabled
    and this always returns True — useful for paper testing / backtests
    where wall-clock time shouldn't block entries.
    """
    s = get_settings()
    if not s.ENFORCE_MARKET_HOURS:
        return True
    ist = to_ist(now)
    if not _is_trading_day(ist):
        return False
    t = ist.time()
    return _parse_hhmm(s.ENTRY_WINDOW_START_IST) <= t <= _parse_hhmm(s.ENTRY_WINDOW_END_IST)


def square_off_due(now: Optional[datetime] = None) -> bool:
    """True iff intraday positions should be force-squared-off now.

    Fires on a trading day once IST time ≥ SQUARE_OFF_TIME (default
    15:10) and up to the close. The EOD scheduler polls this and calls
    `TradeManager.close_all(reason="EOD_SQUAREOFF")` when it flips True.

    Unlike `is_entry_window`, this is NOT disabled by
    `ENFORCE_MARKET_HOURS` — if you square off, you want it to happen on
    the real clock regardless of the entry gate.
    """
    s = get_settings()
    ist = to_ist(now)
    if not _is_trading_day(ist):
        return False
    t = ist.time()
    return _parse_hhmm(s.SQUARE_OFF_TIME_IST) <= t <= _parse_hhmm(s.MARKET_CLOSE_IST)


def entry_block_reason(now: Optional[datetime] = None) -> Optional[str]:
    """Human-readable reason a new entry is blocked by the clock, or
    None when the entry window is open. Handy for risk logs / audit."""
    s = get_settings()
    if not s.ENFORCE_MARKET_HOURS:
        return None
    if is_entry_window(now):
        return None
    if not is_market_open(now):
        return "market_closed"
    if square_off_due(now):
        return "post_squareoff"
    # Open, but inside an excluded edge (first 15 / last 30 minutes).
    return "outside_entry_window"

File: cpp/src/market_clock.cpp
#include "tb/market_clock.hpp"

#include "tb/config.hpp"

namespace tb {
namespace {

// Settings already validates the "HH:MM" format, so this is a thin split.
int parse_hhmm(const std::string& v) {
  const auto colon = v.find(':');
  if (colon == std::string::npos) return 0;
  return std::stoi(v.substr(0, colon)) * 60 + std::stoi(v.substr(colon + 1));
}

bool within(int t, const std::string& lo, const std::string& hi) {
  return parse_hhmm(lo) <= t && t <= parse_hhmm(hi);  // inclusive, as in Python
}

}  // namespace

IstTime to_ist(SysTime now) {
  const auto ist = now + kIstOffset;
  const auto day = std::chrono::floor<std::chrono::days>(ist);
  const std::chrono::weekday wd{day};
  IstTime out;
  out.day = day;
  out.minutes = static_cast<int>(std::chrono::duration_cast<std::chrono::minutes>(ist - day).count());
  // Python's weekday(): Monday=0 .. Friday=4 are trading days.
  out.trading_day = wd.iso_encoding() <= 5;
  return out;
}

bool is_market_open(SysTime now) {
  const IstTime t = to_ist(now);
  if (!t.trading_day) return false;
  const Settings& s = settings();
  return within(t.minutes, s.MARKET_OPEN_IST, s.MARKET_CLOSE_IST);
}

bool is_entry_window(SysTime now) {
  const Settings& s = settings();
  // Disabled gate: paper testing / backtests must not be blocked by the clock.
  if (!s.ENFORCE_MARKET_HOURS) return true;
  const IstTime t = to_ist(now);
  if (!t.trading_day) return false;
  return within(t.minutes, s.ENTRY_WINDOW_START_IST, s.ENTRY_WINDOW_END_IST);
}

bool square_off_due(SysTime now) {
  // Deliberately NOT gated by ENFORCE_MARKET_HOURS: if you square off, it
  // happens on the real clock regardless of the entry gate (invariant I3).
  const IstTime t = to_ist(now);
  if (!t.trading_day) return false;
  const Settings& s = settings();
  return within(t.minutes, s.SQUARE_OFF_TIME_IST, s.MARKET_CLOSE_IST);
}

std::optional<std::string> entry_block_reason(SysTime now) {
  if (!settings().ENFORCE_MARKET_HOURS) return std::nullopt;
  if (is_entry_window(now)) return std::nullopt;
  if (!is_market_open(now)) return "market_closed";
  if (square_off_due(now)) return "post_squareoff";
  // Open, but inside an excluded edge (first 15 / last 30 minutes).
  return "outside_entry_window";
}

}  // namespace tb
